File: app/src/main/java/com/app/ralaunch/utils/ArchiveExtractor.java
package com.app.ralaunch.utils;

import android.content.Context;
import com.app.ralaunch.utils.AppLogger;

import java.io.*;
import java.util.zip.GZIPInputStream;

/**
 * 通用归档文件解压工具
 * 
 * 支持 tar.gz 格式的解压
 */
public class ArchiveExtractor {
    private static final String TAG = "ArchiveExtractor";
    
    /**
     * 解压 tar.gz 文件到目标目录
     * 
     * @param archiveFile tar.gz 文件
     * @param targetDir 目标目录
     * @param stripPrefix 要移除的前缀路径（例如 "usr/lib/" 或 null）
     * @return 解压的文件数量
     * @throws Exception 解压失败时抛出异常
     */
    public static int extractTarGz(File archiveFile, File targetDir, String stripPrefix) throws Exception {
        try (FileInputStream fis = new FileInputStream(archiveFile);
             BufferedInputStream bis = new BufferedInputStream(fis);
             GZIPInputStream gzipIn = new GZIPInputStream(bis);
             org.apache.commons.compress.archivers.tar.TarArchiveInputStream tarIn = 
                 new org.apache.commons.compress.archivers.tar.TarArchiveInputStream(gzipIn)) {
            
            org.apache.commons.compress.archivers.tar.TarArchiveEntry entry;
            int processedFiles = 0;
            
            while ((entry = tarIn.getNextTarEntry()) != null) {
                String entryName = entry.getName();
                
                // 跳过根目录和空条目
                if (entryName.isEmpty() || entryName.equals(".") || entryName.equals("..")) {
                    continue;
                }
                
                // 移除前导的 ./
                if (entryName.startsWith("./")) {
                    entryName = entryName.substring(2);
                }
                
                // 移除指定的前缀
                if (stripPrefix != null && !stripPrefix.isEmpty()) {
                    // 处理 ./usr/lib/ 或 usr/lib/ 前缀
                    if (entryName.startsWith("./" + stripPrefix)) {
                        entryName = entryName.substring(2 + stripPrefix.length());
                    } else if (entryName.startsWith(stripPrefix)) {
                        entryName = entryName.substring(stripPrefix.length());
                    } else if (entryName.contains(stripPrefix)) {
                        // 如果前缀在路径中间，提取前缀之后的部分
                        int idx = entryName.indexOf(stripPrefix);
                        entryName = entryName.substring(idx + stripPrefix.length());
                    } else {
                        // 如果没有找到前缀，跳过不相关的文件（仅当指定了前缀时）
                        // 对于 box64-x64-libs，同时支持 x86_64 和 i386 架构
                        if (stripPrefix.equals("usr/lib/") && 
                            !entryName.contains("box64-x86_64-linux-gnu/") && 
                            !entryName.contains("box64-i386-linux-gnu/")) {
                            continue;
                        }
                        // 对于 SteamCMD，如果没有找到 steamcmd/ 前缀，可能是其他文件，跳过
                        if (stripPrefix.equals("steamcmd/") && !entryName.startsWith("steamcmd/")) {
                            continue;
                        }
                    }
                }
                
                if (entryName.isEmpty()) {
                    continue;
                }
                
                File targetFile = new File(targetDir, entryName);
                
                // 安全检查：防止路径遍历攻击
                String canonicalDestPath = targetDir.getCanonicalPath();
                String canonicalEntryPath = targetFile.getCanonicalPath();
                if (!canonicalEntryPath.startsWith(canonicalDestPath + File.separator) && 
                    !canonicalEntryPath.equals(canonicalDestPath)) {
                    AppLogger.warn(TAG, "跳过不安全的路径: " + entryName);
                    continue;
                }
                
                if (entry.isDirectory()) {
                    if (!targetFile.exists()) {
                        targetFile.mkdirs();
                    }
                } else {
                    // 创建父目录
                    File parent = targetFile.getParentFile();
                    if (parent != null && !parent.exists()) {
                        parent.mkdirs();
                    }
                    
                    // 写入文件
                    try (FileOutputStream fos = new FileOutputStream(targetFile);
                         BufferedOutputStream bos = new BufferedOutputStream(fos)) {
                        byte[] buffer = new byte[8192];
                        int bytesRead;
                        while ((bytesRead = tarIn.read(buffer)) != -1) {
                            bos.write(buffer, 0, bytesRead);
                        }
                    }
                    
                    // 设置权限（如果是可执行文件）
                    if ((entry.getMode() & 0100) != 0) {
                        targetFile.setExecutable(true, false);
                    }
                    
                    // 设置读取权限
                    targetFile.setReadable(true, false);
                }
                
                processedFiles++;
            }
            
            AppLogger.info(TAG, "Extracted " + processedFiles + " files from tar.gz");
            return processedFiles;
            
        } catch (Exception e) {
            AppLogger.error(TAG, "Failed to extract tar.gz", e);
            throw e;
        }
    }
    
    /**
     * 从 assets 复制文件到临时文件
     * 
     * @param context Android 上下文
     * @param assetFileName assets 中的文件名
     * @param targetFile 目标临时文件
     * @throws IOException IO 错误
     */
    public static void copyAssetToFile(Context context, String assetFileName, File targetFile) throws IOException {
        try (InputStream is = context.getAssets().open(assetFileName);
             FileOutputStream fos = new FileOutputStream(targetFile);
             BufferedOutputStream bos = new BufferedOutputStream(fos)) {
            
            byte[] buffer = new byte[8192];
            int bytesRead;
            while ((bytesRead = is.read(buffer)) != -1) {
                bos.write(buffer, 0, bytesRead);
            }
        }
    }
    
    /**
     * 递归删除目录
     * 
     * @param directory 要删除的目录
     */
    public static void deleteDirectory(File directory) {
        if (directory.exists() && directory.isDirectory()) {
            File[] files = directory.listFiles();
            if (files != null) {
                for (File file : files) {
                    if (file.isDirectory()) {
                        deleteDirectory(file);
                    } else {
                        file.delete();
                    }
                }
            }
            directory.delete();
        }
    }
}

File: app/src/main/java/com/app/ralaunch/core/common/util/ArchiveExtractor.kt
package com.app.ralaunch.core.common.util

import android.content.Context
import org.tukaani.xz.XZInputStream
import java.io.*
import java.util.zip.GZIPInputStream

object ArchiveExtractor {
    private const val TAG = "ArchiveExtractor"
    private const val BUFFER_SIZE = 65536 

    fun interface ProgressCallback {
        fun onProgress(processedFiles: Int, currentFile: String)
    }

    @JvmStatic
    @JvmOverloads
    fun extractTarGz(archiveFile: File, targetDir: File, stripPrefix: String?, callback: ProgressCallback? = null): Int {
        return FileInputStream(archiveFile).use { fis ->
            BufferedInputStream(fis, BUFFER_SIZE).use { bis ->
                GZIPInputStream(bis, BUFFER_SIZE).use { gzipIn ->
                    extractTarEntries(gzipIn, targetDir, stripPrefix, callback)
                }
            }
        }
    }

    @JvmStatic
    @JvmOverloads
    fun extractTarXz(archiveFile: File, targetDir: File, stripPrefix: String?, callback: ProgressCallback? = null): Int {
        return FileInputStream(archiveFile).use { fis ->
            BufferedInputStream(fis, BUFFER_SIZE).use { bis ->
                XZInputStream(bis).use { xzIn ->
                    extractTarEntries(xzIn, targetDir, stripPrefix, callback)
                }
            }
        }
    }

    @JvmStatic
    @JvmOverloads
    fun extractTar(archiveFile: File, targetDir: File, stripPrefix: String?, callback: ProgressCallback? = null): Int {
        return FileInputStream(archiveFile).use { fis ->
            BufferedInputStream(fis, BUFFER_SIZE).use { bis ->
                extractTarEntries(bis, targetDir, stripPrefix, callback)
            }
        }
    }

    private fun extractTarEntries(
        inStream: InputStream, targetDir: File,
        stripPrefix: String?, callback: ProgressCallback?
    ): Int {
        var processedFiles = 0
        var lastCallbackTime = 0L 
        val tarReader = MiniTarReader(inStream)

        while (true) {
            val entry = tarReader.nextEntry() ?: break
            val entryName = normalizeEntryName(entry.name, stripPrefix) ?: continue
            val targetFile = File(targetDir, entryName)

            if (!isPathSafe(targetDir, targetFile)) continue

            when {
                entry.isDirectory -> extractDirectory(targetFile)
                entry.isSymbolicLink -> extractSymlink(targetFile, entry.linkName)
                else -> extractFile(tarReader, targetFile, entry.mode)
            }

            processedFiles++
            
            if (callback != null) {
                val currentTime = System.currentTimeMillis()
                if (currentTime - lastCallbackTime > 100) {
                    callback.onProgress(processedFiles, entryName)
                    lastCallbackTime = currentTime
                }
            }
        }

        return processedFiles
    }

    private fun normalizeEntryName(entryName: String?, stripPrefix: String?): String? {
        if (entryName.isNullOrEmpty() || entryName == "." || entryName == "..") return null

        var name: String = entryName
        if (name.startsWith("./")) name = name.substring(2)

        if (!stripPrefix.isNullOrEmpty()) {
            name = when {
                name.startsWith("./$stripPrefix") -> name.substring(2 + stripPrefix.length)
                name.startsWith(stripPrefix) -> name.substring(stripPrefix.length)
                name.contains(stripPrefix) -> name.substring(name.indexOf(stripPrefix) + stripPrefix.length)
                else -> name
            }
        }

        while (name.startsWith("/") || name.startsWith("\\")) {
            name = name.substring(1)
        }

        return name.takeIf { it.isNotEmpty() }
    }

    private fun isPathSafe(targetDir: File, targetFile: File): Boolean {
        return try {
            val canonicalDestPath = targetDir.canonicalPath
            val canonicalEntryPath = targetFile.canonicalPath
            canonicalEntryPath.startsWith("$canonicalDestPath${File.separator}") || canonicalEntryPath == canonicalDestPath
        } catch (e: IOException) { false }
    }

    private fun extractDirectory(targetFile: File) {
        if (!targetFile.exists()) targetFile.mkdirs()
    }

    private fun extractSymlink(targetFile: File, linkTarget: String) {
        targetFile.parentFile?.takeIf { !it.exists() }?.mkdirs()
        if (targetFile.exists()) targetFile.delete()

        try {
            android.system.Os.symlink(linkTarget, targetFile.absolutePath)
        } catch (e: Exception) {
            targetFile.parentFile?.let { parent ->
                val linkTargetFile = File(parent, linkTarget)
                if (linkTargetFile.exists()) {
                    try {
                        linkTargetFile.copyTo(targetFile, overwrite = true)
                    } catch (copyEx: Exception) {}
                }
            }
        }
    }

    private fun extractFile(tarReader: MiniTarReader, targetFile: File, mode: Int) {
        targetFile.parentFile?.takeIf { !it.exists() }?.mkdirs()

        FileOutputStream(targetFile).use { fos ->
            BufferedOutputStream(fos, BUFFER_SIZE).use { bos ->
                val buffer = ByteArray(BUFFER_SIZE)
                while (true) {
                    val read = tarReader.readData(buffer)
                    if (read <= 0) break
                    bos.write(buffer, 0, read)
                }
            }
        }

        if ((mode and 0x40) != 0) targetFile.setExecutable(true, false)
        targetFile.setReadable(true, false)
    }

    @JvmStatic
    fun copyAssetToFile(context: Context, assetFileName: String, targetFile: File) {
        context.assets.open(assetFileName).use { inputStream ->
            FileOutputStream(targetFile).use { fos ->
                BufferedOutputStream(fos, BUFFER_SIZE).use { bos ->
                    val buffer = ByteArray(BUFFER_SIZE)
                    var read: Int
                    while (inputStream.read(buffer).also { read = it } != -1) {
                        bos.write(buffer, 0, read)
                    }
                }
            }
        }
    }

    class MiniTarReader(private val inStream: InputStream) {
        private var currentEntrySize: Long = 0
        private var bytesReadForEntry: Long = 0
        private val headerBuffer = ByteArray(512)

        class TarEntry(
            val name: String,
            val size: Long,
            val typeFlag: Char,
            val linkName: String,
            val mode: Int
        ) {
            val isDirectory: Boolean get() = typeFlag == '5' || name.endsWith("/")
            val isSymbolicLink: Boolean get() = typeFlag == '2'
        }

        fun nextEntry(): TarEntry? {
            val padding = (512 - (bytesReadForEntry % 512)) % 512
            if (padding > 0L) skipFully(padding)

            val read = readFully(headerBuffer)
            if (read < 512) return null
            if (headerBuffer.all { it == 0.toByte() }) return null

            var parsedName = parseString(0, 100)
            var realName: String? = null

            if (parsedName == "././@LongLink") {
                val nameSize = parseOctal(124, 12).toInt()
                val nameBytes = ByteArray(nameSize)
                readFully(nameBytes)
                realName = String(nameBytes).trim('\u0000')

                val namePadding = (512 - (nameSize % 512)) % 512
                if (namePadding > 0) skipFully(namePadding.toLong())

                readFully(headerBuffer)
                parsedName = parseString(0, 100)
            }

            val prefix = parseString(345, 155)
            val finalName = realName ?: if (prefix.isNotEmpty()) "$prefix/$parsedName" else parsedName

            currentEntrySize = parseOctal(124, 12)
            val typeFlag = headerBuffer[156].toInt().toChar()
            val linkName = parseString(157, 100)
            val mode = parseOctal(100, 8).toInt()

            bytesReadForEntry = 0
            return TarEntry(finalName, currentEntrySize, typeFlag, linkName, mode)
        }

        fun readData(buffer: ByteArray): Int {
            if (bytesReadForEntry >= currentEntrySize) return -1
            val toRead = minOf(buffer.size.toLong(), currentEntrySize - bytesReadForEntry).toInt()
            val read = inStream.read(buffer, 0, toRead)
            if (read > 0) bytesReadForEntry += read
            return read
        }

        private fun parseString(offset: Int, length: Int): String {
            var end = offset
            val limit = offset + length
            while (end < limit && headerBuffer[end] != 0.toByte()) end++
            return String(headerBuffer, offset, end - offset).trim()
        }

        private fun parseOctal(offset: Int, length: Int): Long {
            val str = String(headerBuffer, offset, length).trim('\u0000', ' ')
            return if (str.isNotEmpty()) str.toLongOrNull(8) ?: 0L else 0L
        }

        private fun readFully(buffer: ByteArray): Int {
            var totalRead = 0
            while (totalRead < buffer.size) {
                val read = inStream.read(buffer, totalRead, buffer.size - totalRead)
                if (read == -1) break
                totalRead += read
            }
            return totalRead
        }

        private fun skipFully(n: Long) {
            var totalSkipped = 0L
            while (totalSkipped < n) {
                val skipped = inStream.skip(n - totalSkipped)
                if (skipped == 0L) {
                    if (inStream.read() == -1) break else totalSkipped++
                } else {
                    totalSkipped += skipped
                }
            }
        }
    }
}
